pkg/resources/user_password_policy_attachment.go

```diff
@@ -0,0 +1,169 @@
+			StateContext: func(ctx context.Context, d *schema.ResourceData, m interface{}) ([]*schema.ResourceData, error) {
+				parts := strings.Split(d.Id(), helpers.IDDelimiter)
+				if len(parts) != 4 {
+					return nil, fmt.Errorf("id should be in the format 'database|schema|password_policy|user_name|roles', but I got '%s'", d.Id())
+				}
+				passwordPolicyDatabase := sdk.NewAccountIdentifierFromFullyQualifiedName(parts[0])
+				passwordPolicySchema := sdk.NewAccountIdentifierFromFullyQualifiedName(parts[1])
+				passwordPolicyName := sdk.NewAccountIdentifierFromFullyQualifiedName(parts[2])
+				userName := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(parts[3])
+				if err := d.Set("password_policy_database", passwordPolicyDatabase.Name()); err != nil {
+					return nil, err
+				}
+				if err := d.Set("password_policy_schema", passwordPolicySchema.Name()); err != nil {
+					return nil, err
+				}
+				if err := d.Set("password_policy_name", passwordPolicyName.Name()); err != nil {
+					return nil, err
+				}
+				if err := d.Set("user_name", userName.Name()); err != nil {
+					return nil, err
+				}
+				return []*schema.ResourceData{d}, nil
+			},
+		},
+	}
+}
+
+func CreateUserPasswordPolicyAttachment(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
+
+	userName := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(d.Get("user_name").(string))
+	passwordPolicyDatabase := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_database").(string))
+	passwordPolicySchema := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_schema").(string))
+	passwordPolicyName := sdk.NewAccountIdentifierFromFullyQualifiedName(d.Get("password_policy_name").(string))
+	passwordPolicy := sdk.NewSchemaObjectIdentifier(passwordPolicyDatabase.Name(), passwordPolicySchema.Name(), passwordPolicyName.Name())
+
+	err := client.Users.Alter(ctx, userName, &sdk.AlterUserOptions{
+		Set: &sdk.UserSet{
+			PasswordPolicy: passwordPolicy,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_database", passwordPolicyDatabase.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_schema", passwordPolicySchema.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("password_policy_name", passwordPolicyName.Name()); err != nil {
+		return err
+	}
+	if err := d.Set("user_name", helpers.EncodeSnowflakeID(userName)); err != nil {
+		return err
+	}
+	d.SetId(fmt.Sprintf(`%s|%s`, helpers.EncodeSnowflakeID(passwordPolicy), helpers.EncodeSnowflakeID(userName)))
+
+	return ReadUserPasswordPolicyAttachment(d, meta)
+}
+
+func ReadUserPasswordPolicyAttachment(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	ctx := context.Background()
+	userName := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(d.Get("user_name").(string))
```

Comment:
So, to remove custom import and use ImportStatePassthroughContext you need to parse the `d.ID()` to get password policy identifier and user name. `d.Get()` cannot be used in Read operation.

Reply:
I meant that if we use `ImportStatePassthroughContext` (so Read as Import), then Read cannot use `d.Get()`, because Import functionality wouldn't work. Using "Read as Import" requires parsing `d.ID()` to get all needed information for the resource and not rely on the `d.Get()` parameters which are not set yet. If we go with the custom import it can stay as is, but if we go with `ImportStatePassthroughContext` as the Import option then `userName` should be parsed from the `d.ID()`. Reference: https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2307#discussion_r1476262478

Reply:
Ah, I see, now I understand your comment. Thanks for the clarification :) Everything changed :)